src/profile-logic/process-profile.js

```diff
@@ -1052,56 +1056,195 @@ export function processProfile(
+    // and just replace it with a `null` value. We filter out the null values
+    // inside `serializeProfile` function.
+    return null;
+  }
+
+  // We iterate all the markers and remove/change data depending on the PII
+  // status.
+  const markersToDelete = [];
+  if (
+    PIIToBeRemoved.networkUrls ||
+    PIIToBeRemoved.screenshots === true ||
+    (Array.isArray(PIIToBeRemoved.screenshots) &&
+      PIIToBeRemoved.screenshots.length > 0)
+  ) {
+    for (let i = 0; i < thread.markers.length; i++) {
+      const currentMarker = thread.markers.data[i];
+
+      // Remove the all network URLs if user wants to remove them.
+      if (
+        PIIToBeRemoved.networkUrls &&
+        currentMarker &&
+        currentMarker.type &&
+        currentMarker.type === 'Network'
+      ) {
+        // Remove the URI fields from marker payload.
+        currentMarker.URI = '';
+        currentMarker.RedirectURI = '';
+        // Strip the URL from the marker name
+        const stringIndex = thread.markers.name[i];
+        stringArray[stringIndex] = stringArray[stringIndex].replace(/:.*/, '');
+      }
+
+      // If PIIToBeRemoved.screenshots is true, this means that we need to
+      // remove all screenshots, if it's an array of thread IDs, that means we
+      // should only delete screenshots from the given threads.
+      if (
+        (PIIToBeRemoved.screenshots === true ||
+          (Array.isArray(PIIToBeRemoved.screenshots) &&
+            PIIToBeRemoved.screenshots.includes(threadIndex))) &&
+        currentMarker &&
+        currentMarker.type &&
+        currentMarker.type === 'CompositorScreenshot'
+      ) {
+        const urlIndex = currentMarker.url;
+        stringArray[urlIndex] = '';
+        markersToDelete.push(i);
+      }
+    }
+  }
+
+  // After iterating (or not iterating at all) the markers, if we have some
+  // markers we want to delete or user wants to delete the full time range,
+  // reconstruct the marker table and samples table without unwanted information.
+  if (markersToDelete.length > 0 || PIIToBeRemoved.fullTimeRange !== null) {
+    const oldMarkers = thread.markers;
+    const newMarkerTable = getEmptyRawMarkerTable();
+    const addCurrentMarker = (index: number) => {
+      if (markersToDelete.includes(index)) {
+        return;
+      }
+      newMarkerTable.name.push(oldMarkers.name[index]);
+      newMarkerTable.time.push(oldMarkers.time[index]);
+      newMarkerTable.data.push(oldMarkers.data[index]);
+      newMarkerTable.length++;
+    };
+
+    // If user wants to remove full time range, filter all the markers and
+    // samples accordingly.
+    if (PIIToBeRemoved.fullTimeRange !== null) {
+      const { start, end } = PIIToBeRemoved.fullTimeRange;
+      const filteredMarkerIndexIter = filterRawMarkerTableIndexesToRange(
+        oldMarkers,
+        start,
+        end
+      );
+
+      for (const index of filteredMarkerIndexIter) {
```

Comment:
Ok, now I see the other reason for using the generator. Using a generator and `for of` is very terse code for the implementation. As a genuine question, is this implementation worth the magic? I'd at least add more comments for context on what's going on under the hood.

Reply:
On my side I'm not opposed to using a generator, it seems to be the good tool here, and this avoids to loop twice. (maybe we should use the tool more in other places too)

Reply:
Yeah, I really like generators and I think it's a pretty handy feature. It's not common since it's fairly new but probably it's gonna be used more and more in the future.